server/chat-plugins/chatlog.ts

```diff
@@ -166,10 +165,69 @@ export const LogReader = new class {
 		const prevMonth = new Date(new Date(`${month}-15`).getTime() - 30 * DAY);
 		return prevMonth.toISOString().slice(0, 7);
 	}
-
 	today() {
 		return Chat.toTimestamp(new Date()).slice(0, 10);
 	}
+	compareDates(a: string, b: string) {
+		const aDate = new Date(a).getTime();
+		const bDate = new Date(b).getTime();
+		return bDate - aDate;
+	}
+	isDate(item: string) {
+		return !isNaN(new Date(item).getTime());
+	}
+	async getBattleLog(tier: ID, number: number) {
+		// binary search!
+		const months = (await FS('logs').readdir())
+			.filter(f => this.isDate(f) && FS(`logs/${f}/${tier}`).existsSync())
+			.sort(this.compareDates);
+		if (!months.length) return;
+
+		const getBattleNum = (battleName: string) => Number(battleName.split('-')[1].slice(0, -9));
+
+		const getMonthRange = (month: string) => {
+			const files = FS(`logs/${month}/${tier}/`).readdirSync();
+			const firstDay = files[0];
+			const firstPath = FS(`logs/${month}/${tier}/${firstDay}`);
+			const firstFiles = firstPath.readdirSync();
+			if (files.length === 1) {
+				const num = getBattleNum(firstFiles[0]);
+				return [num, num];
+			}
+			const lastDay = files[files.length - 1];
```

Comment:
The order of the files is not guaranteed; you have to sort them before grabbing the last one is different from grabbing a random one.